sync/election/leader_manager.go

```diff
@@ -0,0 +1,358 @@
+	conf.ID += uuid.NewString()[:8] // append a unique suffix to the `Id`
+	return &DlockLeaderManager{
+		Config:      conf,
+		lockMan:     dlm,
+		electionKey: elecKey,
+	}
+}
+
+func (l *DlockLeaderManager) Identity() string {
+	return l.ID
+}
+
+// Await blocks until being elected as leader or context canceled.
+func (l *DlockLeaderManager) Await(ctx context.Context) bool {
+	timer := time.NewTimer(0)
+	defer timer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return false
+		case <-timer.C:
+			if atomic.LoadInt32(&l.electionStatus) == StatusElected &&
+				atomic.LoadInt64(&l.lockExpiryTime) > time.Now().Unix() {
+				return true
+			}
+
+			timer.Reset(l.Retry)
+		}
+	}
+}
+
+// Extend extends leadership lease.
+func (l *DlockLeaderManager) Extend(ctx context.Context) bool {
+	if atomic.LoadInt32(&l.electionStatus) == StatusElected {
+		// acquire the lock to extend the lease
+		return l.acquireLock(ctx) == nil
+	}
+
+	return false
+}
+
+// Campaign starts the election process, which will run in a goroutine until contex canceled.
+func (l *DlockLeaderManager) Campaign(ctx context.Context) {
+	newCtx, cancel := context.WithCancel(ctx)
+	l.cancel = cancel
+
+	logger := logutil.NewErrorTolerantLogger(logutil.DefaultETConfig)
+	etlog := func(err error) {
+		logger.Log(
+			logrus.WithFields(logrus.Fields{
+				"electionKey": l.electionKey,
+				"leaderID":    l.ID,
+			}), err, "Leader election campaign error",
+		)
+	}
+
+	for {
+		// retry to acquire lock at first
+		err := l.retryLock(newCtx)
+		etlog(err)
+
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+
+			l.onError(newCtx, err)
+			time.Sleep(l.Retry)
+			continue
+		}
+
+		// if we got here we are the leader
+		l.onElected(newCtx)
+
+		// renew the lock lease
+		err = l.renewLock(newCtx)
+		etlog(err)
+
+		// if we got here we are not the leader anymore
+		l.onOusted(newCtx)
+
+		if err != nil {
+			if errors.Is(err, context.Canceled) {
+				return
+			}
+
+			l.onError(newCtx, err)
+		}
+	}
+}
+
+// Stop stops the leader election process and resigns from the leadership if appliable.
+func (l *DlockLeaderManager) Stop(ctx context.Context) error {
+	if l.cancel != nil {
+		l.cancel()
+	}
+
+	return l.lockMan.Release(ctx, l.lockIntent())
+}
+
+// retryLock consistently retries to acquire a lock until success.
+func (l *DlockLeaderManager) retryLock(ctx context.Context) error {
+	retryTimer := time.NewTimer(0)
+	defer retryTimer.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-retryTimer.C:
+			err := l.acquireLock(ctx)
+			if errors.Is(err, dlock.ErrLockAcquisitionFailed) {
+				retryTimer.Reset(l.Retry)
+				continue
+			}
+
+			return err
+		}
+	}
+}
+
+// renewLock consistently renews a lock lease until any error.
+func (l *DlockLeaderManager) renewLock(ctx context.Context) error {
+	renewTimer := time.NewTimer(l.Renew)
```

Comment:
This function is similar to "retryLock", consider extracting a generic function with appropriate parameters.

Reply:
I don't think it's feasible since `retryLock` and `renewLock` have different logic. `retryLock` will loop to acquire a distributed lock until success while `renewLock` will loop to acquire a distributed lock until failure.